node/src/neighborhood/node_location.rs

```diff
@@ -0,0 +1,208 @@
+            match country {
+                Some(country) => Some(NodeLocation {
+                    country_code: country.iso3166.to_string(),
+                    free_world_bit: country.free_world,
+                }),
+                None => None,
+            }
+        }
+        None => None,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::neighborhood::gossip::GossipBuilder;
+    use crate::neighborhood::node_location::{get_node_location, NodeLocation};
+    use crate::neighborhood::node_record::{NodeRecord, NodeRecordMetadata};
+    use crate::sub_lib::node_addr::NodeAddr;
+    use crate::test_utils::neighborhood_test_utils::{
+        db_from_node, make_node_record, pick_country_code_record,
+    };
+    use std::net::{IpAddr, Ipv4Addr};
+
+    #[test]
+    fn test_node_location() {
+        let node_location =
+            get_node_location(Some(IpAddr::V4(Ipv4Addr::new(125, 125, 125, 1)))).unwrap();
+
+        assert_eq!(node_location.country_code, "CN");
+        assert_eq!(node_location.free_world_bit, false);
+    }
+
+    #[test]
+    fn construct_node_record_metadata_with_free_world_bit() {
+        let mut metadata = NodeRecordMetadata::new();
+        metadata.node_location_opt = get_node_location(Some(IpAddr::V4(Ipv4Addr::new(1, 1, 1, 1))));
+        assert_eq!(
+            metadata.node_location_opt.as_ref().unwrap(),
+            &NodeLocation {
+                country_code: "AU".to_string(),
+                free_world_bit: true
+            }
+        );
+        assert_eq!(
+            metadata.node_location_opt.as_ref().unwrap().free_world_bit,
+            true
+        );
+        assert_eq!(
+            metadata.node_location_opt.as_ref().unwrap().country_code,
+            "AU"
+        );
+    }
+
+    #[test]
+    fn construct_node_record_for_test() {
+        let mut node_record = make_node_record(1111, true);
+        let country_record = pick_country_code_record(3333);
+        node_record.metadata.node_location_opt = Some(NodeLocation {
+            country_code: country_record.1.clone(),
+            free_world_bit: country_record.2,
+        });
+        node_record.metadata.node_addr_opt =
+            Some(NodeAddr::new(&country_record.0, &[8000 % 10000]));
+        node_record.inner.country_code = country_record.1;
+
+        assert_eq!(
+            node_record.metadata.node_location_opt,
+            Some(NodeLocation {
+                country_code: "AU".to_string(),
+                free_world_bit: true
+            })
+        )
+    }
+
+    #[test]
+    fn create_gossip_node_with_addr_and_country_code_reveal_results_in_node_with_addr_adn_free_world_bit(
+    ) {
+        let mut node = make_node_record(2222, true);
+        let country_record = pick_country_code_record(2222);
+        println!("node: {:?}", node);
+        node.metadata.node_location_opt = Some(NodeLocation {
+            country_code: country_record.1.clone(),
+            free_world_bit: country_record.2,
```

Comment:
Please learn yourself to use the more revealing decoupled tuple with named inner args, not to refer to them by indexes. Here it is:

```Rust
 let (country_code, free_world_bit) = pick_country_code_record(2222);
```